### ISSUE TITLE
refactor iphymv

### ISSUE THREAD

**Issue by trel** [irods/irods#3490]
- [x] master
- [x] 4-2-stable
----------

We're considering removing `iphymv` as an iCommand.

Marking `iphymv` as deprecated in 4.3 (updated July 2019: 4.2.x) would be the first step.

A much more consistent way to migrate files across storage technologies is to use replication (`irepl`) and trimming (`itrim`).

Are there any existing use cases that would break or no longer be possible if `iphymv` were to go away?

**Comment by TD-er** [comment]
It is not really a use case which will break when iphymv is deprecated, but irepl could use some enhancement/option to remove the old replica after a successful replication.
Now one has to determine what replication number the old one had and if that was in a compound resource, there are 2, etc.
Now I think of it, there is a number of new features which would make irepl and itrim more useful. (--purgec for example), but I will open a new issue for that if there isn't one yet.

**Comment by tedgin** [comment]
I often use `iphymv`.  Unlike `irepl`+`itrim`, the replica number of a file doesn't change when it is moved to a new location. At the moment, I use the replica number as part of a file's identity.  Replica 0 is the original upload.  Replica 1 is the remote replica every data object is supposed to have.  Any replicas with number higher than 1 are user initiated replicas. Not having iphymv would make it hard for me to decommission resource servers.

Besides, its nice to have some symmetry between zones and grids.
icp <--> irepl
irm <--> itrim
imv <--> iphymv

**Comment by trel** [comment]
@tedgin This is good feedback - thanks.

We are hoping that decommissioning a server never depends on the specifics of replica numbers.

**Comment by tedgin** [comment]
It doesn't directly. I should have said that without iphmv, it makes it hard to decommission a resource server and still maintain our replica numbering scheme.

**Comment by jasoncoposky** [comment]
I feel that this scheme perhaps is overloading the replica numbers as additional metadata describing the resources on which they reside.  Could this not be managed as actual AVUs within the catalog?

**Comment by kript** [comment]
At present our resource server decommissioning/migrating relies on us syncing the filesystem to a new server, then using `iadmin modresc X` ( inc modrescdatapath), et al to update the resource to the new server. Its worked, but until we're on 4.2.x thats a worrying proposition due to the sheer number of objects to update. We can't use `irepl` and `itrim` for this since we can't target a resource within a compound tree. If we could use `iphymove`, that would allow us to drain gradually. Always open to a better approach, though.. Should/does `iphymove` allow targeting a resource within a compound resource tree?

**Comment by tedgin** [comment]
We could use metadata. A better solution would be to just make my scripts smarter so that they use the locations of the replicas along with number of replicas instead of overloading the replica number.

I would still like to keep iphymv for the symmetry reason as well as the use case @kript  mentions.

**Comment by trel** [comment]
@tedgin I'm not sure I understand your symmetry comment/scenario "between zones and grids".

Can you explain what you mean a bit more?

**Comment by trel** [comment]
@kript Yes, you can target a resource within a resource hierarchy...

Snip of `iphymv -h`:
```
Note that if the source or destination resource is within a hierarchy,
a full resource hierarchy (semicolon delimited) must be specified:
  iphymv -S demoResc -R 'pt1;r1;s1' /tempZone/home/alice/test.txt
```

From https://github.com/irods/irods/issues/3491

**Comment by kript** [comment]
So, can I check that this should work; 
`iphymv -S pt1;r2;s1'  -R 'pt1;r1;s1' /tempZone/home/alice/test.txt`
I.e. moving a file within a resource hierarchy? Or is having the source as the root sufficient?

Also, I _think_ @tedgin was referring to (from his example)  inter-zone one uses icp, intra-zone, aka grid, one uses irepl..

**Comment by tedgin** [comment]
That's not really what I was talking about with symmetry.  Now that I think more deeply about it, the supposed symmetry between grid level, i.e. files, directories and resource servers, and zone level, i.e. data objects, collections, and resources, doesn't really exist.  I'm sorry for my ramblings.

I would still like to keep `iphymv`, but I can live with `irepl` + `itrim`.

**Comment by kript** [comment]
@tedgin Can I ask, do you use composite resource trees and are you running 4.1.X or 4.2 with your use case? We're looking at solutions for this ourself, and would be interesting to know what works for our setup...

**Comment by tedgin** [comment]
We do use composite resource trees.  Since we run 4.1.10, `iphymv` isn't able to transfer data from a leaf storage resource to another leaf storage resource. This means we have to move the storage resource out of the composite resource and then do move the files from the storage one back into the composite.  Unfortunately, `iadmin rmchildfromresc` is too slow for us.  It would take a few months to complete.  I end up doing surgery to the database.  Here's a script that defines our surgery.

```bash
# This script has never been run. It's a fabrication from terminal and postgres histories. 
# There are likely typos.

readonly RescHier="$1"

readonly Store="${RescHier##*;}"
readonly CoordHier="${RescHier%;*}"
readonly Coord="${CoordHier##*;}"

iadmin modresc "$Store" status down

# Wait for current operations on storage resource to cease
while [[ "$connCnt" -ne 1 ]]  # The ips call causes 1 connection
do
  # The first line in an ips report identifies the server, so remove it
  connCnt=$(ips -R "$Store" | tail --lines +2 | wc --lines)
done

readonly CoordSqlList=$(sed "s/\([^;]*\)/\'\1\'/g;s/;/,/g" <<<"$CoordHier")

psql ICAT <<EOSQL
BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ;

UPDATE r_data_main SET resc_name = '$Store', resc_hier = '$Store' WHERE resc_hier = '$RescHier';

UPDATE r_resc_main SET resc_parent = NULL WHERE resc_name = '$Store';

-- The outer REGEXP_REPLACE removes any resulting leading or trailing ;
UPDATE r_resc_main 
SET resc_children = REGEXP_REPLACE(REGEXP_REPLACE(resc_children, ';?$Store;?', ';'), '(^;|;\$)', '')
WHERE resc_name = '$Coord';

UPDATE r_resc_main
SET
  resc_objcount = resc_objcount - (SELECT resc_objcount FROM r_resc_main WHERE resc_name = '$Store')
WHERE resc_name IN ($CoordSqlList);

COMMIT;
EOSQL

iadmin modresc "$Store" status up
~ 
```

We use `iphymv` to actually do the migration, but since it crawls the catalog and checks to see it can move each file individually to see if it should be moved, it is too slow for us.  It would likely take six months to drain one our servers.  We have a script for this, see https://github.com/cyverse/irods-adm/blob/master/phymv.  It finds all of the files that need to be transferred, and then it passes them in batches to multiple iphymv processes running in parallel. This scripts takes a few weeks to run when we decomission a server with  160 TB (~ 10 million files) of data on it.

**Comment by alanking** [comment]
At the moment, all of the use cases above appear to be solved with existing functionality in 4.2.x as tree surgery can be performed quickly and therefore is not a blocker for decommissioning for hardware, and we recommend against specific replica numbers being part of your data management policy. We still think a combination of `irepl` and `itrim` is sufficient.

Given this, we are now considering `iphymv` for deprecation in 4.2.x and for removal in 4.3.x.

**Comment by tedgin** [comment]
The improvements in 4.2 improve the performance of `iadmin rmchildfromresc`, but they don't improve the performance of `iphymv`, `irepl`, and `itrim`, as far as I know.  Since `iphymv` would take six months to drain one of our servers, from my experience, `irepl` would take just as long, and `itrim` isn't much faster. This means that `irepl` + `itrim` would take, maybe nine months instead of six months. Three months savings is significant, but six months is still too long. 

What is needed is something like the automated ingest tool, but for decommissioning a resource server.  For instance, if the resource server belongs to a random resource, this new tool would move all of the replicas on the resource server to one of the other servers belong to the random resource.  If it could do this without interrupting service that would be a very big bonus.

**Comment by kript** [comment]
> What is needed is something like the automated ingest tool, but for decommissioning a resource server. For instance, if the resource server belongs to a random resource, this new tool would move all of the replicas on the resource server to one of the other servers belong to the random resource. If it could do this without interrupting service that would be a very big bonus.

👍 👍 👍 s/very big bonus/dang near _required_ at the kinds of capacities that would need this functionality/
@tedgin has neatly summed up what would signifcantly improve our migration between hardware.

Yes please! Won't someone think of the ~children~ yaks!

**Comment by donsizemore** [comment]
VMWare calls this "maintenance mode" +1

**Comment by trel** [comment]
Moving that much data around serially is going to be slow.  The automated ingest tool 'fixes' this by adding parallelism on the client side.  We could build an external tool to manage the parallelism for this decommissioning/migration scenario (very specific), or we could begin to internally leverage the irodsDelayServer for a variety of recursive or parallel-friendly operations.  That conversation is one we welcome.

Regardless, `iphymv` no longer serves an otherwise-unserved purpose moving forward -- it's a shortcut that punctures the abstractions that iRODS works hard to maintain.

Changing this milestone to 4.2.x for deprecation.

**Comment by kript** [comment]
> We could build an external tool to manage the parallelism for this decommissioning/migration scenario (very specific), or we could begin to internally leverage the irodsDelayServer for a variety of recursive or parallel-friendly operations. That conversation is one we welcome.

:+1: How do we go about kicking off the conversation? Is it a working group, or something more lightweight? Discussion at the next TWG?

**Comment by trel** [comment]
I've put it on the agenda for the next TWG... we can also schedule a call before then to do some specific requirements gathering.

**Comment by trel** [comment]
From the September TWG...

- Discussion
  - tool/rule for decommissioning storage resource
    - need to replace a serial iphymv (deprecated in 4.2)
    - mark to-be-deprecated as read-only via passthru
      - perhaps only a downvote, since updating an existing replica is still a write
    - parallel replication to elsewhere, then trimming
    - possible batch database update rather than per-data-object

Do we not have all the tools we need to do this cleanly already?   It's just not fast because the tools are all serial?

If rebalance were to be parallelized, or if `rebalance` grew a 'background mode' so that it leveraged the rewritten and newly parallel `irodsDelayServer`... then we're much closer?

- Tree surgery would allow the fast/easy removal/addition of old/new resources
- Rebalance will get them into the state you need them in
- `itrim` would still need to become more parallel... or the new `iunreg` https://github.com/irods/irods/issues/4506

**Comment by kript** [comment]
So if I understand correctly, the suggested approach would be;

1. move the to be deprecated resource out of the composite tree
2. `rebalance` the tree to add another replica, potentially running in the background and parallelised
3. remove the now unwanted third replica from the deprecated resource with itrim
4. remove the now empty resource
5. profit!

That seems workable for us I think, but we would like some kind of instrumentation/visibility around the process to see how far the rebalance had got, and a way of logging what itrim was actually doing (dry run vs verbose?), and a way of tuning the parallelism count.

**Comment by trel** [comment]
Yes, that is correct.   You may need to switch your step 1 and 2, depending on whether there are separate good replicas (some people keep three live replicas, rather than two) in the tree-to-be-rebalanced to serve as the source replicas of any rebalancing.

Your wishlist at the bottom...
- Rebalance Visibility - could be provided by a well-crafted query... `select count(DATA_ID) where DATA_RESC_ID = x` where x is the being-populated-resource-id.   The remaining work could be found by comparing this total to the total of a well-populated sibling.

- `itrim` Logging - `--dryrun`, and `-v`, `-V` already exist... can discuss separately as to whether they are good enough to capture what you're interested in.

- Parallelism Tuning - if using the irodsDelayServer, already available via `maximum_number_of_concurrent_rule_engine_server_processes` (which will be renamed as part of https://github.com/irods/irods/issues/3399)

**Comment by tedgin** [comment]
I mostly like the proposal, but it is problematic for us for a couple of reasons.

1. Extra replicas break our current polices that control the number of replicas that can exist.
2. Temporarily creating new root resource by tree surgery complicates our per resource rule logic. We need to make sure the new temporary root resource is combined with its former parent in our rules for things like computing quotas, migrating replicas when a data object move results in it being subjected to different rules, etc.

**Comment by tedgin** [comment]
`iadmin rebalance` wouldn't creating another replica when there aren't multiple replicas in the first place, e.g. in situations where asynchronous replication is used. Would there be something like a parallelized `irepl`?

**Comment by trel** [comment]
I'm not familiar enough with your replica and quota policies to imagine a solution at this time.  

Moving resources in and out of a replication resource (coupled with rebalance) should allow enough control to meet your use cases.  Let's schedule a brainstorming call to make sure we consider the corner cases.


----

Sure, we can consider a parallelized / background `irepl`.

I think, generally, leveraging the `irodsDelayServer` will be how we approach parallelization within the server altogether... we just need to figure out how to allow `irepl`/`itrim`/etc to request parallel work, but actually block until the work is done... so that they can still be scripted against and assume a return code means the work is complete.

**Comment by trel** [comment]
After offline discussion with @tedgin, we have considered renaming this issue from **deprecate `iphymv`** to **refactor `iphymv`** so that it uses the new intermediate replicas, goes through the same API calls as everything else (reducing codepaths), and leverage the new Query Processor Library (https://github.com/irods/irods/issues/4369).  This will be more robust (allow for restarts, consistent error messages), but will also allow the preservation use case of moving a replica of a data object to different storage while maintaining its system metadata (timestamps, replica number, etc.).

This change will also mean that we will not remove `iphymv` in 4.3.0 (https://github.com/irods/irods/issues/4482).

**Comment by kript** [comment]
👍 As part of this refactor, can we also ask the following to be taken into consideration please?
#4508 
#4212 (this is key for us)
#3281

**Comment by trel** [comment]
definitely.  bumping those for visibility.  thanks.

**Comment by mcv21** [comment]
Can I now add #5360 to the iphymv feature list, please? :)

**Comment by alanking** [comment]
Note: Need documentation for this *expected behavior*, too...

```
$ iphymv -S archive_resource -R another_resource foo
remote addresses: 172.17.0.2 ERROR: phymvUtil: phymv error for /tempZone/home/rods/foo, status = -1814000 status = -1814000 DIRECT_ARCHIVE_ACCESS
Level 0: DIRECT_ARCHIVE_ACCESS: unable to open /tempZone/home/rods/foo for unlink
```

**Comment by alanking** [comment]
Okay, time for a summary of the things that will (and will not) be addressed in this issue...

Of the issues mentioned in the thread above:
- #4508
    - Tests will be written for `iphymv` as it interacts with resource hierarchies, both with up and down statuses
    - The `iphymv` portion of the issue will be checked off
    - This *will not* be closed as part of the effort to refactor `iphymv`
- #4212 
    - The issue as originally written is resolved in master and 4-2-stable. However, the discussion in the comments has turned into the classic "extra replica" problem, which will be solved by logical locking (#3848)
    - This *will not* be closed as part of the effort to refactor `iphymv`
- #3281
    - A workaround for the performance bottleneck of a recursive `iphymv` has been proposed in the issue
    - The iRODS CLI of the future will interact with a new and better phymv which will address this. Recursive iteration is currently done on the client side. We are not going to fix the icommand.
    - This *will not* be closed as part of the effort to refactor `iphymv`
- #5360
    - An expansive test suite has been written for various scenarios with `iphymv` and error messages are now bubbled up to the client explaining more specifically what happened in the event of failure
    - This only entails adding messages to the client rError stack, which will be printed out to the user
    - This *will* be closed as part of the effort to refactor `iphymv`

From comment https://github.com/irods/irods/issues/3490#issuecomment-537006600:
> After offline discussion with @tedgin, we have considered renaming this issue from **deprecate `iphymv`** to **refactor `iphymv`** so that it uses the new intermediate replicas, goes through the same API calls as everything else (reducing codepaths), and leverage the new Query Processor Library (#4369). This will be more robust (allow for restarts, consistent error messages), but will also allow the preservation use case of moving a replica of a data object to different storage while maintaining its system metadata (timestamps, replica number, etc.).
> 
> This change will also mean that we will not remove `iphymv` in 4.3.0 (#4482).

Breaking this down...
> uses the new intermediate replicas, goes through the same API calls as everything else (reducing codepaths)

These are one and the same and have been implemented as of the closure of #4343.

> leverage the new Query Processor Library (#4369)

This is very similar if not identical to the issue raised in #3281. So, this will not be addressed in this issue as listed above.

As noted in my previous comment (https://github.com/irods/irods/issues/3490#issuecomment-765556055), we will revisit the documentation for `iphymv` to reflect the new contract it will have with the server.

I think that covers everything. Let me know if you see something I missed, esteemed participants.

**Comment by kript** [comment]
This looks very comprehensive thanks.

**Comment by alanking** [comment]
This will be closed along with #5360. The comment above details the work remaining to be done for `iphymv` as captured in other issues. The marked-down resource behavior will remain a part of the task of #4508 as it will not be practical to tackle the issue for a single API implementation.

**Comment by trel** [comment]
double checkboxed.
